core/services/chainlink/application.go

```diff
@@ -165,6 +152,28 @@ func NewApplication(config *orm.Config, ethClient eth.Client, advisoryLocker pos
 		jobORM         = job.NewORM(store.ORM.DB, store.Config, pipelineORM, eventBroadcaster, advisoryLocker)
 	)
 
+	store.NotifyNewEthTx = ethBroadcaster
+	pendingConnectionResumer := newPendingConnectionResumer(runManager)
+
+	appCallbackLock := &sync.Mutex{}
+	appCallbackLock.Lock()
```

Comment:
Feels weird that the lock leaked out to here.

Why not keep the locking inside the headtracker where the callbacks are executed?

Reply:
@samsondav This is a different lock - it's just to block the goroutine at the end of the `NewApplication()` constructor function, which runs the NewApplication callbacks. We used to add the application level callbacks as callbacks to the head tracker (which was incredibly hacky), and forced this ordering in the constructor where you had to create the app without the head tracker, instantiate the headTracker and then assign it to the app later. This prevents any other service from "consuming" the head tracker, which we now need in the job `Delegate()` function.

This is the best I can explain it haha happy to discuss.